crypto/evp/evp_lib.c

```diff
@@ -1046,3 +1048,35 @@ int EVP_PKEY_CTX_get_group_name(EVP_PKEY_CTX *ctx, char *name, size_t namelen)
         return -1;
     return 1;
 }
+
+EVP_PKEY *EVP_keygen(OSSL_LIB_CTX *libctx, const char *name,
+                      const char *propq, OSSL_PARAM *params)
+{
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_from_name(libctx, name, propq);
+
+    if (ctx != NULL
+            && EVP_PKEY_keygen_init(ctx) > 0
+            && EVP_PKEY_CTX_set_params(ctx, params))
+        (void)EVP_PKEY_gen(ctx, &pkey);
+
+    EVP_PKEY_CTX_free(ctx);
+    return pkey;
+}
+
+EVP_PKEY *EVP_Q_keygen(OSSL_LIB_CTX *libctx, const char *type,
```

Comment:
I'm wondering if one of these two APIs abse on this function prototype might be more friendly and future proof:

```c
EVP_PKEY *EVP_Q_keygen(OSSL_LIB_CTX *libctx,
                       const char *type,
                       const char *propq,
                       ...)
```

The first idea being that the `type` determines the varargs component.  Thus, e.g., RSA would have an `unsigned int` argument specifying the key size and the wrapper could be:
```c
#define EVP_rsa_keygen(libctx, propq, bits) \
    EVP_Q_keygen(libctx, "RSA", propq, (unsigned int)(0 + bits))
```

EC would take a string vararg which could be macroed:
```c
#define EVP_ec_keygen(libctx, propq, curve) \
    EVP_Q_keygen(libctx, "EC", propq, curve)
```

We'd retain some semblance of type checking via the macros and wouldn't be limited when an algorithm needs two arguments or a floating point number.


My other thought is to have name/value pairs in the varargs part.  These continue until a NULL name is encountered which can be checked by some compilers via the `sentinel` attribute.

RSA would thus be:
```c
#define EVP_rsa_keygen(libctx, propq, bits) \
    EVP_Q_keygen(libctx, "RSA", propq, "size", (unsigned int)(0 + bits), NULL)
```

and EC:
```c
#define EVP_ec_keygen(libctx, propq, curve) \
    EVP_Q_keygen(libctx, "EC", propq, "curve", curve, NULL)
```

Once we know the type, the settable params become available and they can be used to determine the type of the argument following the name (there would need to be some restrictions).  We could also, e.g. allow "params" to be followed by an OSSL_PARAM array argument.

Reply:
Interesting thoughts.
Although this PR already introduces the fully general one-shot keygen function
```
EVP_PKEY *EVP_keygen(OSSL_LIB_CTX *libctx, const char *name,
                     const char *propq, OSSL_PARAM *params);
```
I like the idea of making `EVP_Q_keygen()` (or `EVP_PKEY_Q_gen()`, as it is meanwhile called) more elegant and somewhat more flexible. So changed it to
```
EVP_PKEY *EVP_PKEY_Q_gen(OSSL_LIB_CTX *libctx, const char *propq,
                          const char *type, ...);
```
and defined the two convenience macros as
```
# define EVP_RSA_gen(bits) \
    EVP_PKEY_Q_gen(NULL, NULL, "RSA", (unsigned int)(0 + (bits)))
# define EVP_EC_gen(curve) \
    EVP_PKEY_Q_gen(NULL, NULL, "EC", (char *)(strstr(curve, "")))

```
where I added `strstr(curve, "")` in order to provide type checking also for the string variadic arg.

Reply:
Hmm `EVP_keygen` should be probably named `EVP_PKEY_gen_with_params` or maybe even better `EVP_PKEY_gen_from_params`.

Reply:
Using `strstr(curve, "")` is clever.  Another possibly would be `strchr(curve, *curve)`.  I don't see much difference between the two, so this isn't a request to change.  My initial thought for checking the string type was `curve ""` which doesn't work for pointers so I skipped it.  I was in a rush so didn't dwell over other possibilities.

I agree with @t8m that the general function should start with `EVP_PKEY_`.

Reply:
> Using `strstr(curve, "")` is clever. Another possibly would be `strchr(curve, *curve)`. I don't see much difference between the two, so this isn't a request to change. My initial thought for checking the string type was `curve ""` which doesn't work for pointers so I skipped it. I was in a rush so didn't dwell over other possibilities.

I checked also other possibilities and found that  `strstr(curve, "")` is best because, as opposed to alternatives, it is linear in the `curve` parameter, so macro expansion does not cause problems in case the actual argument filled in for `curve` has side effects.

Reply:
> Hmm EVP_keygen should be probably named EVP_PKEY_gen_with_params or maybe even better EVP_PKEY_gen_from_params.

> I agree with @t8m that the general function should start with `EVP_PKEY_`.

I've changed the name of `EVP_keygen()` to `EVP_PKEY_gen()` (rather than a more complicated name) for brevity and consistency with `EVP_PKEY_Q_gen()`.
To this end I renamed `EVP_PKEY_gen()`, which had been introduced for 3.0 earlier, to  `EVP_PKEY_generate()` (this more low-level function is anyway not used often).

Reply:
EVP_PKEY_gen() is truly unfortunate name for this quick generation function and I realize that names proposed by myself above were not good either. The reason is it is not clear from the name whether a key is generated or domain parameters are generated. With the EVP_PKEY_Q_gen (which should be actually EVP_PKEY_Q_keygen unless you plan to somehow support generating domain parameters) - is there actually any reason to have a public one-shot keygen function that uses the OSSL_PARAMS? I'd just drop it. There is IMO too much choice. You have the EVP_PKEY_Q_keygen for the simple stuff and the existing functions for the rest.

And I am fine with renaming EVP_PKEY_gen() to  EVP_PKEY_generate(). IMO the EVP_PKEY_gen() name is ugly.

Reply:
Ok, made `EVP_PKEY_gen()` static (as `evp_pkey_keygen()`)
and renamed `EVP_PKEY_Q_gen()` to `EVP_PKEY_Q_keygen()`.